compaction/compaction_manager.cc

```diff
@@ -351,15 +362,12 @@ future<sstables::compaction_result> compaction_task_executor::compact_sstables(s
         auto sst = t.make_sstable();
         return sst;
     };
-    descriptor.replacer = [this, &t, release_exhausted] (sstables::compaction_completion_desc desc) {
```

Comment:
yeah, the proposed change has smaller code churn. but i think it'd be simpler to reason about if we register / deregister in the single place using the same machinery. the existing logic is already segmented. what we can is, IMHO, to consolidate it for better readability.

Reply:
I think there might be another reason we may need to follow up with the suggestion above.

Old sstables must be only removed from compacting_sstable_registration after on_compaction_completion() call.

Whereas new sstables must be added to compacting_sstable_registration before the on_compaction_completion() call. That's because new sstables added to table's sstable set can be picked by compaction before we had a chance to register them.

Reply:
thanks i see. in other ~~words~~ diagram:

good
---

```mermaid
sequenceDiagram
    participant sst as SSTable
    participant ms as SSTable set
    participant r as compaction registry
    participant c as yet another compaction
    sst ->>+ r: register me
    Note over ms: sst not found in the main set
    sst ->>+ ms: add me to the set
    c --x ms: sst found, but it is already registered
    ms ->>- sst: remove me from main set
    Note over ms: sst not found in the set
    r ->>- sst: deregister me
```
bad
---

```mermaid
sequenceDiagram
    participant sst as SSTable
    participant ms as SSTable set
    participant r as compaction registry
    participant c as yet another compaction
    sst ->>+ ms: add me to main set
    c ->> ms: sst in the set *and* available to be compacted
    sst ->>+ r: register me
    c --x ms: sst found, but it is already registered
    r ->>- sst: deregister me
    c ->> ms: sst in the set *and* available to be compacted
    ms ->>- sst: remove me from main set
```